### DISCUSSION TITLE
If / Then - Statements, How can I achieve this functionally speaking? If  value == xx, do this, else, do that

### DISCUSSION THREAD

**Discussion by jeffward01** [vkhorikov/CSharpFunctionalExtensions#477] [Q&A] [answered]
Hello!

I am just trying to figure out what the functional equivalent of a `if` and `else` statement is. I am sure I can create my own extension methods for this, but there must be something in the library that I am not aware of.  here is my scenario:

**Scenario**

I am trying to call a method, the method has (2) overloads:

```csharp

void MarkForUpdate(DateTime);

void MarkForUpdate<TKey>(TKey, DateTime)
 where TKey : IComparable<TKey>, IEquatable<TKey>;

```

Key is usually a type of `Guid`, so all I want to do is perform  something like below. Here is how I would do this action the **OOP** old way of doing things (not-functional):

```csharp

var _myDateProvider = DateTime.UtcNow;

if (appUserId.IsNullOrDefault())
{
    MarkUpdate(_myDateProvider);
}
else
{
    MarkUpdate(appUserId, _myDateProvider);
}

```

I would love to achieve this functionally, i'm just not sure of the API methods to do so.  

Can anyone provide any suggestions please?

Thanks

**Comment by rutkowskit** [comment, marked as answer]
There are at least 2 options.

Option 1:
```csharp
        Result.Success(_myDateProvider)
            .TapIf(appUserId.IsNullOrDefault(), MarkUpdate)
            .TapIf(appUserId.IsNullOrDefault() is false, dp => MarkUpdate(appUserId.Value, dp));
```
Option 2:
       
```csharp
        Result.SuccessIf(appUserId.IsNullOrDefault(), "appUserId not set")            
            .Tap(() => MarkUpdate(_myDateProvider))
            .TapError(() => MarkUpdate(appUserId.Value, _myDateProvider));    
```

**Reply by jeffward01** [reply]
Thanks for this!  I really appreciate it!

I think that I need to get used to is the Result.Success or the “error (string)” variable. 

In OOP it’s just like “if this, then do that” — and in functional, the thought process of using a Result.Success, and even pairing it with an error message seems so foreign to me. Such as I struggle to grasp why “If this, then that” requires an error message, or a success. 

Both routes in “if this, then that” are technically a success… but I also understand that functional programming is about 2 routes - success and error. 

Thanks for the examples it helps grasp things, hopefully others will appreciate it also
